cpp/src/worker.cpp

```diff
@@ -340,14 +390,34 @@ void Worker::stopProgressThread()
     stopProgressThreadNoWarn();
 }
 
+bool Worker::isProgressThreadRunning() { return _progressThread != nullptr; }
+
 size_t Worker::cancelInflightRequests()
 {
+  size_t canceled = 0;
+
   auto inflightRequestsToCancel = std::make_shared<InflightRequests>();
   {
     std::lock_guard<std::mutex> lock(_inflightRequestsMutex);
     std::swap(_inflightRequestsToCancel, inflightRequestsToCancel);
   }
-  return inflightRequestsToCancel->cancelAll();
+
+  if (isProgressThreadRunning()) {
+    utils::CallbackNotifier callbackNotifierPre{false};
+    registerGenericPre([&callbackNotifierPre, &canceled, &inflightRequestsToCancel]() {
+      canceled = inflightRequestsToCancel->cancelAll();
+      callbackNotifierPre.store(true);
+    });
+    callbackNotifierPre.wait([](auto flag) { return flag; });
+
+    utils::CallbackNotifier callbackNotifierPost{false};
+    registerGenericPost([&callbackNotifierPost]() { callbackNotifierPost.store(true); });
+    callbackNotifierPost.wait([](auto flag) { return flag; });
```

Comment:
Yes, same as above we must ensure the work has actually completed to avoid things like destroying resources while the progress thread is still running those.